Add support for "controller" keyword for configuring routes controllers

voronkovich opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

This PR adds a syntax sugar for configuring routes controllers in more user-friendly way by using a `controller` keyword.

```yaml
# Current syntax
# It looks strange and exposes Symfony's internals
blog_show:
    path:     /blog/{slug}
    defaults: { _controller: AppBundle:Blog:show }

# Suggested syntax
blog_show:
    path:       /blog/{slug}
    controller: AppBundle:Blog:show
```
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<routes xmlns="http://symfony.com/schema/routing" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">

    <!-- Current syntax -->
    <route id="blog_list" path="/blog">
        <default key="_controller">AppBundle:Blog:list</default>
    </route>

    <!-- Suggested syntax -->
    <route id="blog_list" path="/blog" controller="AppBundle:Blog:list" />
</routes>
```

voronkovich committed:
[Routing] Use "controller" keyword for configuring routes controllers

ogizanagi commented on src/Symfony/Component/Routing/Tests/Fixtures/controller_aware.yml:
```diff
@@ -0,0 +1,9 @@
+
+app_blog:
+    path: /blog
+    controller: AppBundle:Homepage:show
```
Any reason for this not to throw, but let the defaults override it?
IMHO, it'd be better to throw an exception when both are set, in order to avoid some potential WTF moments.

voronkovich replied:
@ogizanagi, I think you're right. I've changed the code to throw the exception when both controller attribute and _controller default are set.

stof commented on src/Symfony/Component/Routing/Loader/ControllerAwareXmlFileLoader.php:
```diff
@@ -0,0 +1,69 @@
+ *
+ * @author Oleg Voronkovich <oleg-voronkovich@yandex.ru>
+ */
+class ControllerAwareXmlFileLoader extends XmlFileLoader
```
Why using a separate file loader instead of adding the feature in the existing one ? AFAICT, adding the new feature is fully backward compatible.

ogizanagi replied:
I asked myself the same. But I came to the conclusion it's legit, as the controller parameter is "tied" to the HttpKernel usage of the Routing component.
But then, I wonder if this rather belong to the `HttpKernel` component or the `FrameworkBundle` instead?

ogizanagi replied:
However, `RouteCollectionBuilder::add()` also set the controller and is in the component itself...
Eventually, it's probably not worth it to create separate loaders for this.

voronkovich replied:
@stof, The routing component doesn’t know about "controllers", "views", "models" and all things like that. 

> The Routing component maps an HTTP request to a set of configuration variables.

The `controller` keyword is valueless when you use the routing component as a standalone library. So I thinks it's better to extend the existing loaders than change them.

stof commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -27,7 +27,7 @@
 class YamlFileLoader extends FileLoader
 {
-    private static $availableKeys = array(
+    protected static $availableKeys = array(
```
This introduces inheritance extension points, which then have to be maintained forever. And inheritance extension points are the worse to maintain for BC. So -1

javiereguiluz commented:
I'm not sure about this proposal. It's better for DX ... but it introduces another way to configure something that you can already configure it in several ways.

stof commented:
Thus, it might then confuse people about why there is `_controller` in attributes of the Request object

voronkovich commented:
@stof, I think it's not a problem. Most users prefer to configure routes using annotations and don't know about the Request's `_controller` parameter.

fabpot commented:
I like the idea. Back then, I didn't know if we would have more than `_controller`. 5 years later, I think it's a good time to actually recognize that using `_controller` is not so nice and move it to a first class citizen.

nicolas-grekas commented:
Same to me, DX wise, it's way easier to use.
About the implem, I think we should do that in the existing loaders.
I get the argument that the router doesn't know about what a "controller" is, but a loader just maps some input to an output. Adding this case doesn't mean the routing components becomes controller-aware all of a sudden. This will ease the patch and its maintenance.

szymach commented:
Out of curiosity, is there a way to not define a controller and have the route work? Aside from it being just a placeholder, like for example the logout route. If there is no other way to use routes that to map them to controllers, it should just by a first-class citizen.

stof commented:
I totally agree about using the existing loader. It also means that other projects uisng the component without FrameworkBundle (Silex, Drupal, etc...) get the same experience.

voronkovich commented:
@stof, @nicolas-grekas, Okay, maybe you're right. I've made changes to use the existing loaders instead of creating the new ones.

fabpot commented on src/Symfony/Component/Routing/Loader/XmlFileLoader.php:
```diff
@@ -29,6 +30,26 @@ class XmlFileLoader extends FileLoader
+     * A controller's key name (e.g. '_controller').
+     *
+     * @var string
+     */
```
We don't document private properties. This can be removed

fabpot commented on src/Symfony/Component/Routing/Loader/XmlFileLoader.php:
```diff
@@ -29,6 +30,26 @@ class XmlFileLoader extends FileLoader
+    private $controllerKey;
+
+    /**
+     * Constructor.
```
Can eb removed

fabpot commented on src/Symfony/Component/Routing/Loader/XmlFileLoader.php:
```diff
@@ -29,6 +30,26 @@ class XmlFileLoader extends FileLoader
+    /**
+     * Constructor.
+     *
+     * @param FileLocatorInterface $locator       A FileLocatorInterface instance
```
This can be removed as well

fabpot commented on src/Symfony/Component/Routing/Loader/XmlFileLoader.php:
```diff
@@ -229,6 +250,19 @@ private function parseConfigs(\DOMElement $node, $path)
+                throw new \InvalidArgumentException(sprintf(
+                    'The routing file "%s" must not specify both the "controller" attribute and the defaults key "%s" for %s.',
+                    $path, $this->controllerKey, $name
+                ));
```
should be on one line

fabpot commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -28,11 +29,31 @@
+     *
+     * @var mixed
+     */
+    private $controllerKey;
```
phpdocs to be removed

fabpot commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -28,11 +29,31 @@
+    private $controllerKey;
+
+    /**
+     * Constructor.
```
to be removed

fabpot commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -28,11 +29,31 @@
+    /**
+     * Constructor.
+     *
+     * @param FileLocatorInterface $locator       A FileLocatorInterface instance
```
to be removed

fabpot commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -204,5 +233,11 @@ protected function validate($config, $name, $path)
+            throw new \InvalidArgumentException(sprintf(
+                'The routing file "%s" must not specify both the "controller" key and the defaults key "%s" for "%s".',
+                $path, $this->controllerKey, $name
+            ));
```
should be on one line

voronkovich commented:
@fabpot, I've fixed all the CS issues you pointed out.

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/XmlFileLoader.php:
```diff
@@ -28,6 +29,18 @@ class XmlFileLoader extends FileLoader
+    /**
+     * @param string $controllerKey A controller's key name
+     */
+    public function __construct(FileLocatorInterface $locator, $controllerKey = 'controller')
```
why not `_controller` as default?
why do we need this to be configurable?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/routing.xml:
```diff
@@ -26,11 +26,13 @@
         <service id="routing.loader.xml" class="Symfony\Component\Routing\Loader\XmlFileLoader">
             <tag name="routing.loader" />
             <argument type="service" id="file_locator" />
+            <argument>_controller</argument>
```
if `_controller` is going to be the default in the loaders constructors, then this should be removed (same below)

voronkovich commented:
@nicolas-grekas, Fixed. I've removed the `controllerKey` property. Now the `_controller` key is hardcoded in the loaders.

nicolas-grekas commented:
W don't merge PRs with merge commit, please rebase instead.

voronkovich commented:
@nicolas-grekas, sorry, I used the GitHub's web interface to resolve conflict and didn't know that it creates merge commit. I've removed that commit and rebased the PR. Thanks!

nicolas-grekas commented on src/Symfony/Component/Routing/Tests/Loader/YamlFileLoaderTest.php:
```diff
@@ -13,8 +13,8 @@
 
 use PHPUnit\Framework\TestCase;
 use Symfony\Component\Config\FileLocator;
-use Symfony\Component\Routing\Loader\YamlFileLoader;
```
please keep original ordering (that's our policy: alpha order for new items, "as is" for existing ones)

nicolas-grekas commented on src/Symfony/Component/Routing/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Added support for prioritized routing loaders.
  * Add matched and default parameters to redirect responses
+ * Added support for a `controller` keyword for configuring routes controllers in YAML and XML configurations. 
```
route~~s~~ controllers (route with no "s")

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/XmlFileLoader.php:
```diff
@@ -11,11 +11,11 @@
 
 namespace Symfony\Component\Routing\Loader;
 
-use Symfony\Component\Routing\RouteCollection;
```
please keep original ordering

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/XmlFileLoader.php:
```diff
@@ -229,6 +229,16 @@ private function parseConfigs(\DOMElement $node, $path)
 
+        if ($controller = $node->getAttribute('controller')) {
+            if (isset($defaults['_controller'])) {
+                $name = $node->hasAttribute('id') ? sprintf('"%s"', $node->getAttribute('id')) : sprintf('the %s tag', $node->tagName);
```
missing quotes around the 2nd `%s`

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/XmlFileLoader.php:
```diff
@@ -229,6 +229,16 @@ private function parseConfigs(\DOMElement $node, $path)
+            if (isset($defaults['_controller'])) {
+                $name = $node->hasAttribute('id') ? sprintf('"%s"', $node->getAttribute('id')) : sprintf('the %s tag', $node->tagName);
+
+                throw new \InvalidArgumentException(sprintf('The routing file "%s" must not specify both the "controller" attribute and the defaults key "_controller" for %s.', $path, $name));
```
would be great to formulate the message so that $file is last

voronkovich replied:
@nicolas-grekas, I just copied an existing message from here: https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/Routing/Loader/YamlFileLoader.php#L191

sstok replied:
`Route "%s" must must not specify both the "controller" attribute and the defaults key "_controller" in file "%s"'` ?

voronkovich replied:
@sstok, yes, but all other messages have a format with a file path at the beginning. I think we should keep the consistency.

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -11,12 +11,12 @@
 
 namespace Symfony\Component\Routing\Loader;
 
-use Symfony\Component\Routing\RouteCollection;
```
original ordering

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -116,6 +116,10 @@ protected function parseRoute(RouteCollection $collection, $name, array $config,
         $methods = isset($config['methods']) ? $config['methods'] : array();
         $condition = isset($config['condition']) ? $config['condition'] : null;
 
+        if ($controller = isset($config['controller']) ? $config['controller'] : null) {
```
```
if (!empty($config['controller'])) {
    $defaults['_controller'] = $config['controller'];
}

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -141,6 +145,10 @@ protected function parseImport(RouteCollection $collection, array $config, $path
         $schemes = isset($config['schemes']) ? $config['schemes'] : null;
         $methods = isset($config['methods']) ? $config['methods'] : null;
 
+        if ($controller = isset($config['controller']) ? $config['controller'] : null) {
```
see above

nicolas-grekas reviewed (approved):
(failure unrelated)

xabbuh commented on src/Symfony/Component/Routing/Tests/Fixtures/controller/routing.xml:
```diff
@@ -0,0 +1,8 @@
+    xsi:schemaLocation="http://symfony.com/schema/routing
+        http://symfony.com/schema/routing/routing-1.0.xsd">
+
+    <route id="app_homepage" path="/" controller="AppBundle:Homepage:show" />
```
I think we should also have another imported route that does not specify the `controller` attribute.

xabbuh replied:
And what about a case where an imported route makes use of the `_controller` default?

xabbuh commented on src/Symfony/Component/Routing/Tests/Loader/XmlFileLoaderTest.php:
```diff
@@ -287,4 +287,44 @@ public function testNullValuesInMap()
+     * @expectedException \InvalidArgumentException
+     * @expectedExceptionMessageRegExp /The routing file "[^"]*" must not specify both the "controller" attribute and the defaults key "_controller" for the "import" tag/
+     */
+    public function testImportWithOverridedController()
```
Overridden

xabbuh commented on src/Symfony/Component/Routing/Tests/Loader/YamlFileLoaderTest.php:
```diff
@@ -108,4 +108,44 @@ public function testLoadWithResource()
+     * @expectedException \InvalidArgumentException
+     * @expectedExceptionMessageRegExp /The routing file "[^"]*" must not specify both the "controller" key and the defaults key "_controller" for "_static"/
+     */
+    public function testImportWithOverridedController()
```
Overridden

xabbuh commented on src/Symfony/Component/Routing/Tests/Loader/YamlFileLoaderTest.php:
```diff
@@ -108,4 +108,44 @@ public function testLoadWithResource()
+    public function testOverrideControllerInDefaults()
+    {
+        $loader = new YamlFileLoader(new FileLocator(array(__DIR__.'/../Fixtures/controller')));
+        $routeCollection = $loader->load('override_defaults.yml');
```
`$routeCollection` is not used

xabbuh commented on src/Symfony/Component/Routing/Tests/Loader/YamlFileLoaderTest.php:
```diff
@@ -108,4 +108,44 @@ public function testLoadWithResource()
+    public function testImportWithOverridedController()
+    {
+        $loader = new YamlFileLoader(new FileLocator(array(__DIR__.'/../Fixtures/controller')));
+        $routeCollection = $loader->load('import_override_defaults.yml');
```
`$routeCollection` is not used

xabbuh commented on src/Symfony/Component/Routing/Tests/Loader/XmlFileLoaderTest.php:
```diff
@@ -287,4 +287,44 @@ public function testNullValuesInMap()
+    public function testImportWithOverridedController()
+    {
+        $loader = new XmlFileLoader(new FileLocator(array(__DIR__.'/../Fixtures/controller')));
+        $routeCollection = $loader->load('import_override_defaults.xml');
```
`$routeCollection` is not used

xabbuh commented on src/Symfony/Component/Routing/Tests/Loader/XmlFileLoaderTest.php:
```diff
@@ -287,4 +287,44 @@ public function testNullValuesInMap()
+    public function testOverrideControllerInDefaults()
+    {
+        $loader = new XmlFileLoader(new FileLocator(array(__DIR__.'/../Fixtures/controller')));
+        $routeCollection = $loader->load('override_defaults.xml');
```
`$routeCollection` is not used

xabbuh commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -141,6 +145,10 @@ protected function parseImport(RouteCollection $collection, array $config, $path
         $schemes = isset($config['schemes']) ? $config['schemes'] : null;
         $methods = isset($config['methods']) ? $config['methods'] : null;
 
+        if (!empty($config['controller'])) {
```
Why not `isset()` instead?

xabbuh commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -116,6 +116,10 @@ protected function parseRoute(RouteCollection $collection, $name, array $config,
         $methods = isset($config['methods']) ? $config['methods'] : array();
         $condition = isset($config['condition']) ? $config['condition'] : null;
 
+        if (!empty($config['controller'])) {
```
Why not `isset()` instead?

ro0NL commented on src/Symfony/Component/Routing/Loader/YamlFileLoader.php:
```diff
@@ -204,5 +212,8 @@ protected function validate($config, $name, $path)
                 $name, $path
             ));
         }
+        if (isset($config['controller']) && isset($config['defaults']['_controller'])) {
```
can be written as `isset($a, $b)`

javiereguiluz replied:
Sadly you can't. See #22684.

ro0NL replied:
You're right, i've forgot about that. :+1:

ro0NL reviewed (commented):
removed my downvote :) this is solved rather elegant actually :+1:

voronkovich commented:
@xabbuh, thanks for the code review! I've fixed all issues you pointed out and added the missing test cases.

xabbuh commented on src/Symfony/Component/Routing/Tests/Loader/XmlFileLoaderTest.php:
```diff
@@ -287,4 +287,64 @@ public function testNullValuesInMap()
+
+        $route = $routeCollection->get('app_homepage');
+
+        $this->assertSame('FrameworkBundle:Template:template', $route->getDefault('_controller'));
```
here should now have an assertion for the other import route to have the `_controller` attribute to be set to the value configured in the `import.xml` file (same for the `YamlFileLoaderTest`)

voronkovich replied:
@xabbuh, I've added the missing assertions.

xabbuh reviewed (commented):
What about a case where the importing file makes use of the new `controller` option while the imported file makes use of the `_controller` attribute (and vice versa)? Should that be a supported use case?

voronkovich commented:
@xabbuh, I think you're right! I've added the test case you suggested.

nicolas-grekas commented:
Thank you @voronkovich.

nicolas-grekas commented:
@voronkovich can you please open a doc PR/issue?

voronkovich commented:
@nicolas-grekas, the issue is already created (thank you, @HeahDude!): https://github.com/symfony/symfony-docs/issues/8240

Pierstoval commented:
> @szymach commented on 3 Jul
> Out of curiosity, is there a way to not define a controller and have the route work? Aside from it being just a placeholder, like for example the logout route. If there is no other way to use routes that to map them to controllers, it should just by a first-class citizen.

FYI, the `defaults: { _controller: ... }` is not mandatory. It's just "default" values. `_controller` is only used by the kernel to know what callable will be executed to render the response. If you create routes with no controller and handle the responses in kernel listeners instead, it'll be ok.
